modules/core/src/test/java/org/apache/ignite/internal/processors/database/BPlusTreeReplaceRemoveRaceTest.java

```diff
@@ -0,0 +1,421 @@
+import org.apache.ignite.internal.processors.cache.persistence.diagnostic.pagelocktracker.PageLockTrackerManager;
+import org.apache.ignite.internal.processors.cache.persistence.tree.BPlusTree;
+import org.apache.ignite.internal.processors.cache.persistence.tree.io.BPlusIO;
+import org.apache.ignite.internal.processors.cache.persistence.tree.io.BPlusInnerIO;
+import org.apache.ignite.internal.processors.cache.persistence.tree.io.BPlusLeafIO;
+import org.apache.ignite.internal.processors.cache.persistence.tree.io.IOVersions;
+import org.apache.ignite.internal.processors.cache.persistence.tree.io.PageIO;
+import org.apache.ignite.internal.processors.cache.persistence.tree.reuse.ReuseList;
+import org.apache.ignite.internal.processors.failure.FailureProcessor;
+import org.apache.ignite.internal.util.typedef.T2;
+import org.apache.ignite.testframework.GridTestUtils;
+import org.apache.ignite.testframework.junits.GridTestKernalContext;
+import org.apache.ignite.testframework.junits.common.GridCommonAbstractTest;
+import org.junit.Test;
+
+/**
+ * Test is based on {@link BPlusTreeSelfTest} and has a partial copy of its code.
+ */
+public class BPlusTreeReplaceRemoveRaceTest extends GridCommonAbstractTest {
+    /** */
+    private static final short PAIR_INNER_IO = 30000;
+
+    /** */
+    private static final short PAIR_LEAF_IO = 30001;
+
+    /** */
+    protected static final int PAGE_SIZE = 512;
+
+    /** */
+    protected static final long MB = 1024 * 1024;
+
+    /** */
+    private static final int CACHE_ID = 100500;
+
+    /** */
+    protected PageMemory pageMem;
+
+    /** Tracking of locks holding. */
+    private PageLockTrackerManager lockTrackerManager;
+
+    /** {@inheritDoc} */
+    @Override protected void beforeTest() throws Exception {
+        pageMem = createPageMemory();
+
+        lockTrackerManager = new PageLockTrackerManager(log, "testTreeManager");
+
+        lockTrackerManager.start();
+    }
+
+    /** {@inheritDoc} */
+    @Override protected void afterTest() throws Exception {
+        if (pageMem != null)
+            pageMem.stop(true);
+
+        if (lockTrackerManager != null)
+            lockTrackerManager.stop();
+    }
+
+    /**
+     * @return Page memory.
+     */
+    protected PageMemory createPageMemory() {
+        DataRegionConfiguration plcCfg = new DataRegionConfiguration()
+            .setInitialSize(1024 * MB)
+            .setMaxSize(1024 * MB);
+
+        PageMemory pageMem = new PageMemoryNoStoreImpl(log,
+            new UnsafeMemoryProvider(log),
+            PAGE_SIZE,
+            plcCfg,
+            new DataRegionMetricsImpl(plcCfg, new GridTestKernalContext(log())),
+            true);
+
+        pageMem.start();
+
+        return pageMem;
+    }
+
+    /**
+     * @return Allocated meta page ID.
+     * @throws IgniteCheckedException If failed.
+     */
+    private FullPageId allocateMetaPage() throws IgniteCheckedException {
```

Comment:
Maybe reanme to **allocateIdxMetaPage**

Reply:
What does it have to do with index?

Reply:
I see that the index partition appears here, but it doesn't seem to matter.

Reply:
Yes, any partition could be used there